dec/decode.go

```diff
@@ -65,3 +67,102 @@ func DecompressBuffer(encodedBuffer []byte, decodedBuffer []byte) ([]byte, error
 func toC(array []byte) *C.uint8_t {
 	return (*C.uint8_t)(unsafe.Pointer(&array[0]))
 }
+
+// Decompresses a Brotli-encoded stream using the io.Reader interface
+type BrotliReader struct {
+	state  C.BrotliState
+	reader io.Reader
+
+	// Internal buffer for compressed data
+	buffer []byte
+
+	availableIn C.size_t
+	nextIn      *C.uint8_t
+	totalOut    C.size_t
+}
+
+// Fill a buffer, p, with the decompressed contents of the stream.
+// Returns the number of bytes read, or an error
+func (r *BrotliReader) Read(p []byte) (int, error) {
+	var err error
+
+	// Prepare arguments
+	maxOutput := len(p)
+	availableOut := C.size_t(maxOutput)
+	nextOut := (*C.uint8_t)(unsafe.Pointer(&p[0]))
+	var read int
+
+	for availableOut > 0 {
+
+		// Read more compressed data
+		if r.availableIn == 0 {
+			var read int
+			if read, err = r.reader.Read(r.buffer); err != nil {
+				return 0, err
+			}
+			r.availableIn = C.size_t(read)
+			r.nextIn = (*C.uint8_t)(unsafe.Pointer(&r.buffer[0]))
+		}
+
+		if r.availableIn > 0 {
+			// Decompress
+			result := C.BrotliDecompressStream(
+				&r.availableIn,
+				&r.nextIn,
+				&availableOut,
+				&nextOut,
+				&r.totalOut,
+				&r.state,
+			)
+
+			read = maxOutput - int(availableOut)
+
+			switch result {
+			case C.BROTLI_RESULT_SUCCESS:
+				break
+			case C.BROTLI_RESULT_NEEDS_MORE_OUTPUT:
+				if read > 0 {
+					return read, nil
+				} else {
+					return 0, errors.New("Brotli decompression error: needs more output buffer")
+				}
+			case C.BROTLI_RESULT_ERROR:
+				return 0, errors.New("Brotli decompression error")
+			case C.BROTLI_RESULT_NEEDS_MORE_INPUT:
+				continue
+			default:
+				return 0, errors.New("Unrecognised Brotli decompression error")
+			}
+		}
+	}
+
+	return read, nil
+}
+
+func (r *BrotliReader) Close() error {
+
+	C.BrotliStateCleanup(&r.state)
+
+	if v, ok := r.reader.(io.Closer); ok {
+		return v.Close()
+	}
+
+	return nil
+}
+
+// Returns a Reader that decompresses the stream from another reader.
+// Ensure that you Close the stream when you are finished in order to clean up the
+// Brotli compression state
+func Decompress(stream io.Reader) io.ReadCloser {
+
+	r := &BrotliReader{
```

Comment:
From my initial experiments, enc and dec need to be separate, as there is too much overlap in the source code to allow it to build into one package. And you're right, being able to exclude the encoder is nice (it's 600kb in the .a file on os x vs 72kb for the decoder), but I feel that the Go community isn't one to care about binary size that much.

You can't easily refer to C data structures from other packages:
> Cgo translates C types into equivalent unexported Go types. Because the translations are unexported, a Go package should not expose C types in its exported API: a C type used in one Go package is different from the same C type used in another.

This means that we would have to expose the access through some wrapper in the enc and dec packages (maybe a type declaration is enough?) which would then allow the structure to be used from the main brotli package.

Regarding API breakage - if I tag the current head as v1 and update the documentation regarding gopkg.in, it wouldn't be too bad for those wanting the new structure to upgrade to v2. I've already broken some backwards compatibility by switching to gopkg in the first place, but the repo's only been published for a week, so hopefully the damage is minimal.

Suggested plan:
 * Merge the stream support
 * Tag as v1
 * Reorganise with internal enc/dec packages which are then used by the main brotli package
 * Tag as v2